### PR TITLE
Issue #11 Docker support

### PR DESCRIPTION
**Pull request by trioletas**
👷 Work in progress:

- [x] Production image
- [x] Test image
- [x] Development mode (OSX/Native)
- [x] Development mode (OSX/Toolbox)
- [x] Development mode (Linux)
- [x] Development mode (Windows/Toolbox)
- [x] Development mode (Windows/Native)
- [ ] Integration with CI tools

### CODE REVIEW

**Comment by trioletas** [comment]
Thanks for looking at the prod image. imma look in to pm2 integration for prod as well - wonder if it makes sense to use it in a container environment.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+WORKDIR /opt/app
+COPY . /opt/app
+
+RUN yarn build && \
+    yarn --production --ignore-scripts --prefer-offline && \
+    yarn cache clean
+
+# Run
+
+ENV NODE_ENV production
```
Comment (jurgisrudaks) [inline Dockerfile]: Woudln't it make the code a little more well structured and more readable, if we would bring all the ENV definitions to the top of the file, in one place?

Comment (trioletas) [reply Dockerfile]: Thanks, I'll try to move this one to the top but the reason why it's here in the first place is that if you set node env to production during install phase, you won't be able to install dev depenendenices required during build.

Comment (trioletas) [reply Dockerfile]: Hey, I've restructured the Dockerfile to use the multi stage builds and now all vars are set in the `run` image at the bottom.

**Comment by trioletas** [comment]
Added the multi stage built support, pm2 shelved for now, will work on dev mode.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+COPY package.json .
+COPY yarn.lock .
+
+RUN yarn --production --ignore-scripts --prefer-offline
+RUN cp -R node_modules production_node_modules
+RUN yarn
+
+# Build
+
+FROM base AS build
```
Comment (jelisejev) [inline Dockerfile]: Why are build and dependencies in different stages?

**Comment by jelisejev** [comment]
@trioletas does adding multi stage builds reduce the size in this particular case?

**Comment by trioletas** [comment]
Multi step builds serve the purpose of simplifying the dockerfile structure by making the usual clean up after yourself routine unnecessary. You can definitely get the same image size without them, but it's not as future proof to do the manual clean up.

File: Dockerfile (hunk 2 above)
Comment (trioletas) [reply Dockerfile]: The idea was to have an image per logical step: base,install,build,(test?),run
It does not serve any practical purpose at the moment.
Does that bother you?

**Comment by trioletas** [comment]
There was a bug in the non-multistage implementation with original node modules folder with all dev dependencies being present in the final image. I've decided to go with this approach to avoid these sort of mishaps with a more maintainable structure

**Comment by trioletas** [comment]
Dev mode is now properly supported with dedicated Dockerfile and docker-compose. Compilation and linting runs in the container, source code directory is attached as a volume mount. I don't see a way how to reuse the productive Dockerfile at the moment. Tested in OS X native and windows docker toolbox, looking for volunteers to test other setups as well.

**Comment by trioletas** [comment]
@aeldar would be nice if you could assist us by testing the development set up on Linux machine, I remember that you have one available.
- install docker engine for Linux
- checkout this branch
- run `docker-compose up`
- change a couple of files on file system and verify if the changes are reflect in a container runtime

thanks in advance!

**Comment by aeldar** [comment]
@trioletas works for me. Checked on the following configuration:
- Linux Mint 18.2 (Ubuntu 16.04 based),
- Docker 17.05
- Docker Compose 1.16.1

Changes in the source files were reflected in a container runtime.

**Comment by trioletas** [comment]
Thanks a lot Eldar !

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+version: '3'
```
Comment (jelisejev) [inline docker-compose.yml]: Wanna add a docker compose file for build as well?

File: Dockerfile (hunk 2 above)
Comment (jelisejev) [reply Dockerfile]: Not sure, do these stages add any performance overhead?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    tty: true
+    build:
+      context: .
+      dockerfile: Dockerfile.dev
+    volumes:
+      - ./src:/opt/app/src
+    ports:
+      - 3001:3001
+    environment:
+      - SHELL=/bin/sh
```
Comment (jelisejev) [inline docker-compose.yml]: How will we pass other env variables to the app?

File: README.md [hunk 5]
```diff
@@ -37,15 +38,21 @@ Koa, GraphQL and TypeScript template project with batteries included.
...
+- Docker
+
+## Install
+
+`$ npm i` or `$ yarn` or `$ docker-compose up`
 
 ## Develop
 
-`$ npm start` or `$ yarn start`
+`$ npm start` or `$ yarn start` or `$ docker-compose up`
```
Comment (jelisejev) [inline README.md]: By default `docker-compose up` does not re-build the image and you might end up running an old cached version. You need to add --build.

File: docker-compose.yml (hunk 4 above)
Comment (trioletas) [reply docker-compose.yml]: '.env'

File: docker-compose.yml (hunk 3 above)
Comment (trioletas) [reply docker-compose.yml]: What's the use case? I've scoped the compose to dev mode only.

File: README.md (hunk 5 above)
Comment (trioletas) [reply README.md]: Fair enough, will add a remark.

File: Dockerfile (hunk 2 above)
Comment (trioletas) [reply Dockerfile]: A bit, probably. Haven't noticed any negative influence on performance so far nor have seen any mention of serious impact elsewhere. I would like to keep it like that for now, this structure helps me reason about what happens on which stage better.

File: docker-compose.yml (hunk 4 above)
Comment (trioletas) [reply docker-compose.yml]: That ok?

**Pull request merged**